File: apps/api/internal/domain/group.go
package domain

import "time"

// GroupVisibility represents whether a group is publicly discoverable or private.
type GroupVisibility string

// GroupVisibility values.
const (
	GroupVisibilityPublic  GroupVisibility = "public"
	GroupVisibilityPrivate GroupVisibility = "private"
)

// GroupMemberRole represents a member's role within a group.
type GroupMemberRole string

// GroupMemberRole values.
const (
	GroupRoleOwner  GroupMemberRole = "owner"
	GroupRoleMember GroupMemberRole = "member"
)

// GroupMemberStatus represents whether a membership is pending approval or active.
type GroupMemberStatus string

// GroupMemberStatus values.
const (
	GroupMemberStatusPending GroupMemberStatus = "pending"
	GroupMemberStatusActive  GroupMemberStatus = "active"
)

// MembershipNone is the sentinel returned for a viewer who is not related to a group.
const MembershipNone = "none"

// Group represents a club or crew created by an owner (armador). Public groups are
// join-by-request (owner approval); private groups are joined with an invite code.
type Group struct {
	ID          string
	OwnerID     string
	Name        string
	Description *string
	PhotoURL    *string
	Visibility  GroupVisibility
	InviteCode  *string
	CreatedAt   time.Time
	UpdatedAt   time.Time

	// Derived fields populated by repository reads (not stored on the groups row).
	MemberCount        int
	PendingCount       int
	MyMembershipStatus string // "none" | "pending" | "active"
	MyRole             string // "" | "owner" | "member"
}

// GroupMember represents a user's membership in a group.
type GroupMember struct {
	GroupID  string
	UserID   string
	Role     GroupMemberRole
	Status   GroupMemberStatus
	JoinedAt time.Time
}

File: apps/mobile/lib/features/groups/domain/entities/group.dart
/// A club or crew. Public groups are join-by-request (owner approval); private
/// groups are joined with an invite code.
class Group {
  const Group({
    required this.id,
    required this.ownerId,
    required this.name,
    required this.visibility,
    this.description,
    this.photoUrl,
    this.inviteCode,
    this.memberCount = 0,
    this.pendingCount = 0,
    this.myMembershipStatus = 'none',
    this.myRole = '',
    this.createdAt,
    this.updatedAt,
  });

  final String id;
  final String ownerId;
  final String name;
  final String? description;
  final String? photoUrl;
  final String visibility; // 'public' | 'private'
  final String? inviteCode; // only present for the owner
  final int memberCount;
  final int pendingCount;
  final String myMembershipStatus; // 'none' | 'pending' | 'active'
  final String myRole; // '' | 'owner' | 'member'
  final DateTime? createdAt;
  final DateTime? updatedAt;

  bool get isPublic => visibility == 'public';
  bool get isOwner => myRole == 'owner';
  bool get isActiveMember => myMembershipStatus == 'active';
  bool get isPending => myMembershipStatus == 'pending';

  Group copyWith({
    String? id,
    String? ownerId,
    String? name,
    String? description,
    String? photoUrl,
    String? visibility,
    String? inviteCode,
    int? memberCount,
    int? pendingCount,
    String? myMembershipStatus,
    String? myRole,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Group(
      id: id ?? this.id,
      ownerId: ownerId ?? this.ownerId,
      name: name ?? this.name,
      description: description ?? this.description,
      photoUrl: photoUrl ?? this.photoUrl,
      visibility: visibility ?? this.visibility,
      inviteCode: inviteCode ?? this.inviteCode,
      memberCount: memberCount ?? this.memberCount,
      pendingCount: pendingCount ?? this.pendingCount,
      myMembershipStatus: myMembershipStatus ?? this.myMembershipStatus,
      myRole: myRole ?? this.myRole,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Group && runtimeType == other.runtimeType && id == other.id;

  @override
  int get hashCode => id.hashCode;
}
